### PR TITLE
fix slice stackoverflow in certain scenarios

### LINKED ISSUE #286: Slice method will Stack Overflow in certain scenarios
```dart
import 'package:collection/collection.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test("slice1", () {
    final List<int> someList = [1, 2, 3, 4, 5];
    final List<int> result = someList.slice(0, 3); // return [1, 2, 3]
    result.slice(0, 1); // Stack Overflow
  });

  test("slice2", () {
    final List<int> someList = [1, 2, 3, 4, 5];
    final ListSlice<int> result = someList.slice(0, 3); // return [1, 2, 3]
    result.slice(0, 1); // normal return [0]
  });

  test("slice3", () {
    final List<int> someList = [1, 2, 3, 4, 5];
    final result = someList.slice(0, 3); // return [1, 2, 3]
    result.slice(0, 1); // normal return [0]
  });
}
```
As above, when using List<int> to receive .slice results again .slice will cause Stack Overflow.

The code looks as expected ListExtensions#slice | ListSlice#slice, but I can't restrict my downstream from receiving .slice results with a List<int> . Is there any good way to solve it.

### PR DESCRIPTION
**Pull request by hcanyz**
fix #286

### CODE REVIEW

**Comment by devoncarew** [comment]
@hcanyz - thanks for the contribution! Can you update the changelog and the pubspec version as per https://github.com/dart-lang/.github/blob/main/PULL_REQUEST_TEMPLATE.md?

File: lib/src/list_extensions.dart [hunk 1]
```diff
@@ -241,7 +241,7 @@ extension ListExtensions<E> on List<E> {
   ListSlice<E> slice(int start, [int? end]) {
     end = RangeError.checkValidRange(start, end, length);
     var self = this;
-    if (self is ListSlice) return self.slice(start, end);
+    if (self is ListSlice<E>) return self.slice(start, end);
```
Comment (lrhn) [inline lib/src/list_extensions.dart:244]: (This is a constant foot-gun. I'll add another notch to the issue suggesting we promote to `ListSlice<E>` here.)

File: lib/src/list_extensions.dart [hunk 2]
```diff
@@ -404,7 +404,7 @@ class ListSlice<E> extends ListBase<E> {
   /// ```
   ListSlice<E> slice(int start, [int? end]) {
     end = RangeError.checkValidRange(start, end, length);
-    return ListSlice._(_initialSize, source, start + start, end - start);
+    return ListSlice._(_initialSize, source, this.start + start, end - start);
```
Comment (lrhn) [inline lib/src/list_extensions.dart:407]: This was just bad. Sorry :cry:

**Review by lrhn — APPROVED** [review]
Good fix, thank you.

As @devoncarew said, upping the version in the SDK (just a minor version increment is enough) and adding an entry in CHANGELOG.md saying what that version does (fixes bug in `List.slice` and `ListSlice.slice` methods), should be enough to make it publishable.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,7 @@
+## 1.17.3
+
+* Fixes bug in List.slice and ListSlice.slice methods.
```
Comment (devoncarew) [inline CHANGELOG.md:3]:
```suggestion
* Fixes a bug in the `List.slice` and `ListSlice.slice` methods.
```

**Comment by hcanyz** [comment]
1.18.0-wip has been committed on the master branch, and I don't know how to proceed with submitting this PR. Or you guys come and commit this change : ) .

**Pull request closed without merging**